exciting_environments/cart_pole/cart_pole_env.py

```diff
@@ -276,37 +277,59 @@ def vector_field(t, y, args):
+        ref = self.PhysicalState(deflection=jnp.nan, velocity=jnp.nan, theta=jnp.nan, omega=jnp.nan)
+        return self.State(physical_state=physical_states, PRNGKey=PRNGKey, additions=additions, reference=ref)
 
     @partial(jax.jit, static_argnums=0)
-    def init_state(self):
+    def init_state(self, env_properties, rng: chex.PRNGKey = None, vmap_helper=None):
         """Returns default initial state for all batches."""
-        phys = self.PhysicalState(
-            deflection=jnp.zeros(self.batch_size),
-            velocity=jnp.zeros(self.batch_size),
-            theta=jnp.full(self.batch_size, jnp.pi),
-            omega=jnp.zeros(self.batch_size),
+        if rng is None:
+            phys = self.PhysicalState(
+                deflection=0,
+                velocity=0,
+                theta=jnp.pi,
+                omega=0,
+            )
+            subkey = None
+        else:
+            state_norm = jax.random.uniform(rng, minval=-1, maxval=1, shape=(4,))
+            phys = self.PhysicalState(
+                deflection=state_norm[0] * env_properties.physical_constraints.deflection,
+                velocity=state_norm[1] * env_properties.physical_constraints.velocity,
+                theta=state_norm[2] * env_properties.physical_constraints.theta,
+                omega=state_norm[3] * env_properties.physical_constraints.omega,
+            )
+            key, subkey = jax.random.split(rng)
+        additions = None  # self.Optional(something=jnp.zeros(self.batch_size))
+        ref = self.PhysicalState(deflection=jnp.nan, velocity=jnp.nan, theta=jnp.nan, omega=jnp.nan)
+        return self.State(physical_state=phys, PRNGKey=subkey, additions=additions, reference=ref)
+
+    @partial(jax.jit, static_argnums=0)
+    def vmap_init_state(self, rng: chex.PRNGKey = None):
+        return jax.vmap(self.init_state, in_axes=(self.in_axes_env_properties, 0, 0))(
+            self.env_properties, rng, jnp.ones(self.batch_size)
         )
-        opt = None  # self.Optional(something=jnp.zeros(self.batch_size))
-        return self.State(physical_state=phys, PRNGKey=None, optional=opt)
 
     @partial(jax.jit, static_argnums=0)
-    def default_reward_func(self, obs, action, action_constraints):
+    def generate_reward(self, state, action, env_properties):
         """Returns reward for one batch."""
-        reward = (
-            (0.01 * obs[0]) ** 2
-            + 0.1 * (obs[1]) ** 2
-            + (obs[2]) ** 2
-            + 0.1 * (obs[3]) ** 2
-            + 0.1 * (action[0] / action_constraints.force) ** 2
-        )
+        reward = 0
+        for name in self.control_state:
+            reward += -(
+                (
+                    (getattr(state.physical_state, name) - getattr(state.reference, name))
+                    / (getattr(env_properties.physical_constraints, name)).astype(float)
+                )
+                ** 2
```

Comment:
The angle information (angle of the pendulum) should be transformed to sin and cos here. Otherwise the loss is kind of inconsistent